### ISSUE TITLE
ios_config: multi-line banner

### ISSUE THREAD

**Issue by ajacoutot** [ansible/ansible-modules-core#5318]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

ios_config
##### ANSIBLE VERSION

```
ansible 2.2.0.0 (stable-2.2 bce9bfce51) last updated 2016/10/19 09:43:58 (GMT +200)
  lib/ansible/modules/core: (detached HEAD 730f26d52f) last updated 2016/10/19 09:44:56 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD d51b3d5887) last updated 2016/10/19 09:44:56 (GMT +200)
  config file = /home/ajacoutot/work/ansible.cfg
  configured module search path = Default w/o overrides

```
##### CONFIGURATION

None.
##### OS / ENVIRONMENT

Running on Fedora, managing Cisco 2960 switches.
##### SUMMARY

Trying to configure a 'banner motd' on the device does not work with multi-line.
##### STEPS TO REPRODUCE

Using this file as content source :

```
banner motd #
Welcome to foo
How are you today
#
```

And this task :

```
- name: deploy banner
  ios_config:
    src: "/path/to/banner"
    provider: "{{ cli }}"
```

Ansible fails with :

```
TASK [cisco-ios-config : deploy banner ] **************************
fatal: [sw1]: FAILED! => {"changed": false, "failed": true, "msg": "timeout trying to send command: banner motd #\r"}
    to retry, use: --limit @/home/ajacoutot/work/deploy-conf.retry

PLAY RECAP *********************************************************************
sw1            : ok=18   changed=1    unreachable=0    failed=1 
```

It does work fine with single-line banner.
e.g.

```
banner motd # Welcome to this switch #
```
##### EXPECTED RESULTS

I'd expect multi-line banner to work like single-line do.
##### ACTUAL RESULTS

See above.
Thank you :-)

**Comment by ansibot** [comment]
@privateip, @gundalow, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by gundalow** [comment]
@ajacoutot Thank you for the bug report

/cc @privateip @Qalthos Is this something that needs fixing for 2.2?

**Comment by bdowling** [comment]
Yet another quirk with IOS configs that is not idempotent.  The banner command accepts almost any delim character, % and # are commonly used as long as it is not contained within the text, but when it saves the config it changes it to a Control-C.  When doing a show running it prints it as two characters ^C on the banner line and at then end of the potentially multi-line banner.

Note also that there are no prompts for the banner input lines.  They just need to be sent followed by the ending delimiter character.  e.g.

```
ciscodev(config)#banner login %
Enter TEXT message.  End with the character '%'.

You are logging into $(hostname) on $(line) ($(line-desc)).  Do you have the right access?

Unauthorized users will be skewered and roasted over open flame.
%
ciscodev(config)#
```

(not all devices support the variables either, but ansible could fill in the devicename if desired..)

I was thinking of writing a handler for this case, but I'm not certain I see any examples of special case handlers like this already to template it after.

Does the team have thoughts on how they want to handle cases like this?  Should it be magic "fixups" in ios_config or a special ios_banner module?  Thoughts?

**Comment by bdowling** [comment]
@privateip fwiw, I was toying around with a few things here..  Banner seems a little different than the other command that have various prompts, since in this case just pasting the blurb should be enough to make it work.  I noticed that yaml could have a multi-line entry, so I tried working with | for the banner input.  Then I realized that ios_config was splitting this on newlines no matter what.  So I made a fix to test theory:

```
     if configobjs:
-        commands = dumps(configobjs, 'commands').split('\n')
+#        if want_to_split:
+#          commands = dumps(configobjs, 'commands').split('\n')
+#        else:
+        commands = dumps(configobjs, 'commands')
```

I'm not sure if that "block style" is kept track of anywhere or if another flag might be an option (send_as_one_cmd or somesuch).  I also see dumps() has an output param, but not sure if that helps here.  

Just thought I'd throw this out there since you took this issue ;)

**Example Plays:**

```
- hosts: oneswitch
  gather_facts: no
  connection: local

  tasks:

  - name: set device banner from multi-line entry
    ios_config:
      authorize: yes
      commands: |
        banner motd %
        Welcome..  Please keep in mind..

        Unauthorized users will be skewered and roasted over open flame.

        Have a nice day.
        %
      match: none
      provider: "{{ cli }}"

  - name: set device banner from multi-line entry
    ios_config:
      authorize: yes
      commands: |
        banner login %

        You are logging into {{inventory_hostname}}

        This is a multi-line banner.  Have a great day.
        %
      match: none
      provider: "{{ cli }}"

```

On the other end of this, obviously performing a match is going to be harder since cisco mogrify's the delimiter (% in this case) to ^C ...  I thought perhaps there should be some sort of "match_fixup_regex_sub" that could be performed either against the current config or the command lines before the match is performed?   This feature could come in handy for another condition I noted where cisco does not always return index numbers for access lists, even when you specify odd numbers.  So doing the match ignoring the numbers from the lines would work in that case or matching end_of_lines.  :thought_balloon:

**Comment by privateip** [comment]
I have some working code that I hope to merge by weeks end that will address the banner issue.  The change focuses more on the specific banner use case by extracting from the config and candidate and doing the comparison and remediation independently of the rest of the config.  While this will fix the bug, one of my big items for 2.3 is going to be to clean up netcfg and integration multiline support for these types of issues (it resurfaces for loading thinks like certifications and keys)

**Comment by privateip** [comment]
I just committed a change that should fix the problem.  @ajacoutot could you re-test and confirm the fix?

needs_info

**Comment by ajacoutot** [comment]
@privateip sure thing, but I will not be able to do so before Monday -- I hope that's alright.
Thanks!

**Comment by bdowling** [comment]
@privateip #5494  has a few issues.  breaks ios_config, if match is none, gather_facts is off or the device doesn't have banners:

```
line 287, in run
    banners = diff_banners(want_banners, have_banners)
    UnboundLocalError: local variable 'have_banners' referenced before assignment
```

Other issues:
- does not currently handle 'banner **motd**'  Not sure if there are other banner types missing as well.  This perhaps could be more flexible and just use `banner (\w+)`
- the desired config must be specified with ^C as the delimiter
- when sending config to device it is hard-coded to use @ character as banner cmd delimiter, preventing people who, say, might want to put an admin email address in a banner.
- it appears as though this only works with 'src' file?  Is that correct?  Does not handle use inline, e.g. multiple `lines:` or `lines: |` formatting in playbook?

**Comment by ajacoutot** [comment]
@privateip I can confirm what @bdowling wrote.
I was able to make 'banner login' work with multiline using a template (ios_config), so that's already an improvement though :-)
It'd be nice to have support for 'banner motd' has mentioned as well.
Thank you very much, it's going into the right direction imho.

**Comment by privateip** [comment]
committing additional changes in #5524 to fix the issues noted above.  Two caveats.  

1) Still requires configuring the banner using src argument
2) Using ^C is still required in the candidate config file but you can now configure the delimiter used during configuration changes.  See the `multiline_delimiter`

**Comment by ajacoutot** [comment]
@privateip That works perfectly for my use case.
Thank you _very_ much :-)

**Comment by privateip** [comment]
fixed and verified

**Comment by myref** [comment]
Great work, could you post an example config?

**Comment by Len357** [comment]
I am not able to get banner motd or banner login to work. I am using the dev branch. Any help with a sample config would be great. I have tried replacing ^C in the src file with @. After everything I do I keep getting a timeout message.

dminlen@len-nm-1:~/ansible$ ansible --version
ansible 2.3.0
config file = /home/adminlen/ansible/ansible.cfg
configured module search path = Default w/o overrides
adminlen@len-nm-1:~/ansible$ cat router-login-banner.yml

    hosts: testsitehsrp1
    gather_facts: no
    connection: local

    tasks:
        name: Set motd and alias commands for dsl
        ios_config:
        username: adminlen
        save: yes
        src: /home/adminlen/ansible/router-login-banner-test

adminlen@len-nm-1:~/ansible$ cat router-login-banner-test
banner login @

test login
banner

@

adminlen@len-nm-1:~/ansible$ ansible-playbook -vvv router-login-banner.yml --ask-pass
Using /home/adminlen/ansible/ansible.cfg as config file
SSH password:

PLAYBOOK: router-login-banner.yml *******************************************************************************
1 plays in router-login-banner.yml

PLAY [testsitehsrp1] ********************************************************************************************
META: ran handlers

TASK [Set motd and alias commands for dsl] **********************************************************************
task path: /home/adminlen/ansible/router-login-banner.yml:7
Using module file /usr/local/lib/python2.7/dist-packages/ansible-2.3.0-py2.7.egg/ansible/modules/network/ios/ios_config.py
<10.100.80.2> ESTABLISH LOCAL CONNECTION FOR USER: adminlen
<10.100.80.2> EXEC /bin/sh -c 'echo ~ && sleep 0'
<10.100.80.2> EXEC /bin/sh -c '( umask 77 && mkdir -p "echo /home/adminlen/.ansible/tmp/ansible-tmp-1487961451.38-133078886647534" && echo ansible-tmp-1487961451.38-133078886647534="echo /home/adminlen/.ansible/tmp/ansible-tmp-1487961451.38-133078886647534" ) && sleep 0'
<10.100.80.2> PUT /tmp/tmpIbJY6x TO /home/adminlen/.ansible/tmp/ansible-tmp-1487961451.38-133078886647534/ios_config.py
<10.100.80.2> EXEC /bin/sh -c 'chmod u+x /home/adminlen/.ansible/tmp/ansible-tmp-1487961451.38-133078886647534/ /home/adminlen/.ansible/tmp/ansible-tmp-1487961451.38-133078886647534/ios_config.py && sleep 0'
<10.100.80.2> EXEC /bin/sh -c '/usr/bin/python /home/adminlen/.ansible/tmp/ansible-tmp-1487961451.38-133078886647534/ios_config.py; rm -rf "/home/adminlen/.ansible/tmp/ansible-tmp-1487961451.38-133078886647534/" > /dev/null 2>&1 && sleep 0'
fatal: [testsitehsrp1]: FAILED! => {
"changed": false,
"command": "banner login @",
"failed": true,
"invocation": {
"module_args": {
"after": null,
"auth_pass": null,
"authorize": null,
"backup": false,
"before": null,
"config": null,
"defaults": false,
"force": false,
"host": null,
"lines": null,
"match": "line",
"multiline_delimiter": "@",
"parents": null,
"password": null,
"port": null,
"provider": null,
"replace": "line",
"save": true,
"src": "banner login @ \n\ntest login\nbanner\n\n@\n\n",
"ssh_keyfile": null,
"timeout": null,
"username": "adminlen"
}
},
"msg": "timeout trying to send command: banner login @",
"rc": 1
}
to retry, use: --limit @/home/adminlen/ansible/router-login-banner.retry

PLAY RECAP ******************************************************************************************************
testsitehsrp1 : ok=0 changed=0 unreachable=0 failed=1
